File: code/java/剑指offer13.java
class Solution {
    public boolean judge(int a,int b,int k){
        int a_1 = a/10;
        int a_0 = a%10;
        int b_1 = b/10;
        int b_0 = b%10;
        if(a_1+a_0+b_1+b_0>k) return false;
        return true;
    }

    public int movingCount(int m, int n, int k) {
        Stack<Integer> stack = new Stack<>();
        int[][] used = new int[m][n];
        //起点
        stack.push(0);
        while(!stack.isEmpty()){
            int cur = stack.pop();
            int row = cur/n;
            int col = cur%n;
            used[row][col] = 1;
            if(row>0&&(used[row-1][col]==0&&judge(row-1,col,k)))stack.push((row-1)*n+col);
            if(row<m-1&&(used[row+1][col]==0&&judge(row+1,col,k))) stack.push((row+1)*n+col);
            if(col>0&&(used[row][col-1]==0&&judge(row,col-1,k))) stack.push((row)*n+col-1);
            if(col<n-1&&(used[row][col+1]==0&&judge(row,col+1,k))) stack.push((row)*n+col+1);
        }

        int res = 0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                res+=used[i][j];
            }
        }
        return res;
    }
}

File: code/python/剑指offer13.py
class Solution:
    def movingCount(self, m: int, n: int, k: int) -> int:
        def judge(row,col,k):
            a = row//10
            a_ = row%10
            b = col//10
            b_ = col%10
            if a+a_+b+b_>k:
                return False
            return True
        used = [[False]*n for _ in range(m)]
        stack = [(0,0)]
        while len(stack)>0:
            cur = stack.pop()
            used[cur[0]][cur[1]] = True
            if cur[0]>0 and used[cur[0]-1][cur[1]]==False and judge(cur[0]-1,cur[1],k)==True:
                stack.append((cur[0]-1,cur[1]))
            if cur[0]<m-1 and used[cur[0]+1][cur[1]]==False and judge(cur[0]+1,cur[1],k)==True:
                stack.append((cur[0]+1,cur[1]))
            if cur[1]>0 and used[cur[0]][cur[1]-1]==False and judge(cur[0],cur[1]-1,k)==True:
                stack.append((cur[0],cur[1]-1))
            if cur[1]<n-1 and used[cur[0]][cur[1]+1]==False and judge(cur[0],cur[1]+1,k)==True:
                stack.append((cur[0],cur[1]+1))
        # for i in range(m):
        #     print(used[i])
        return sum([sum(used[i]) for i in range(m)])